File: ios/chezz/Models/Streak.swift
import Foundation

// Day-streak math, kept pure (no Date()/UserDefaults) so it is unit-testable. A "day" is a calendar
// day in the given calendar's timezone; playing any game on consecutive days extends the streak.
enum Streak {
    // The streak after playing at `now`, given the prior count and the last day a game was played.
    static func afterPlay(prev: Int, lastPlayed: Date?, now: Date, calendar: Calendar = .current) -> (count: Int, lastPlayed: Date) {
        let today = calendar.startOfDay(for: now)
        guard let lastPlayed else { return (1, today) }
        let last = calendar.startOfDay(for: lastPlayed)
        if last == today { return (max(prev, 1), today) }          // already counted today
        let gap = calendar.dateComponents([.day], from: last, to: today).day ?? 0
        return (gap == 1 ? prev + 1 : 1, today)                    // consecutive extends; any gap resets
    }

    // The streak to display at `now`: alive if the last play was today or yesterday, else lapsed (0).
    static func current(count: Int, lastPlayed: Date?, now: Date, calendar: Calendar = .current) -> Int {
        guard let lastPlayed, count > 0 else { return 0 }
        let today = calendar.startOfDay(for: now)
        let last = calendar.startOfDay(for: lastPlayed)
        let gap = calendar.dateComponents([.day], from: last, to: today).day ?? 0
        return (0...1).contains(gap) ? count : 0
    }
}

File: server/src/lib/streak.ts
// Consecutive-day play streak, mirroring the iOS Streak logic (Streak.swift). Day boundaries are
// computed in UTC from server time only (never a client-supplied date), so a client cannot inflate
// its streak: afterPlay caps the gain at +1 per server-day and is idempotent within the same day.

const DAY_MS = 86_400_000;

function startOfDayUTC(d: Date): number {
  return Date.UTC(d.getUTCFullYear(), d.getUTCMonth(), d.getUTCDate());
}

// New (count, lastPlayed) after a play event today. lastPlayed is the real instant `now`; day
// boundaries are recomputed on read (currentStreak / the next afterPlay) so we never persist a
// normalized-midnight value that could shift across a day when round-tripped through the DB.
export function afterPlay(
  prev: number,
  lastPlayed: Date | null,
  now: Date,
): { count: number; lastPlayed: Date } {
  const today = startOfDayUTC(now);
  if (!lastPlayed) return { count: 1, lastPlayed: now };
  const last = startOfDayUTC(lastPlayed);
  if (last === today) return { count: Math.max(prev, 1), lastPlayed: now }; // already counted today
  const gap = Math.round((today - last) / DAY_MS);
  return { count: gap === 1 ? prev + 1 : 1, lastPlayed: now };
}

// Displayable streak: lapses to 0 once 2+ calendar days have passed without play.
export function currentStreak(count: number, lastPlayed: Date | null, now: Date): number {
  if (!lastPlayed || count <= 0) return 0;
  const gap = Math.round((startOfDayUTC(now) - startOfDayUTC(lastPlayed)) / DAY_MS);
  return gap >= 0 && gap <= 1 ? count : 0;
}
